### PR TITLE
[DLNA] [NextTrack] [Gapless] Fix 5644 and 5294

### LINKED ISSUE #5294: gaps between tracks ...
Well, how to start, I know the jellyfin crew are excellent so i'm asking here first ....

Albums such as many pink floyd, have no gaps between the tracks. When ripped and I play them back they have gaps, which is bad in this case.

What is the cause of this gap?

I use the following:

1. I rip using fairstars free CD ripper - it seems to work very well
2. I run Jellyfin on my NAS, plenty of ram, reasonable processor and nothing other than Ubuntu running on the NAS
3. I stream to a Naim Uniti 2

I am not missing (as far as i can tell) any of the music in the tracks, they just don't flow seamlessly between one track and the next as they would with the original CD (i.e. there is a noticeable gap)

Is the likely cause Jellyfin? the Naim not pre-fetching the next track (stream)? the way the fairstars rips the tracks (i really hope not - re-ripping all my cd's .. ugh) - but at least an easy solution

The Jellyfin software runs smoothly on a HP mirocserver (Gen 8 , Xeon CPU, upgraded RAM, Unbuntu "Bionic"), no log errors, but Jellyfin does highlight a slow CPU on startup. That said it handles streaming ripped DVD's without any hiccups etc, therefore think CPU speed is an unlikely culprit.

Thanks for any insight you can provide
Dave

### LINKED ISSUE #5644: [DLNA] [Music] Next track command from any DLNA device does not do anything
**Describe the bug**

When playing a list of music tracks on a DLNA device, the next track command works fine in Jellyfin, but it does not do anything on the DLNA device.

**System (please complete the following information):**
 - OS: All
 - Virtualization: Docker and without
 - Clients: DLNA device.
 - Browser: Any.
 - Jellyfin Version: 10.6.4 and 10.7.x.
 - Playback: Any 
 - Installed Plugins: Any
 - Reverse Proxy: No
 - Base URL: None
 - Networking: Host
 - Storage: local

**To Reproduce**

1. Go to your favorite Music Library.
2. Cast (Play to) your favorite DLNA device (mine is Sonos or Marantz)
3. Launch a play list using PlayAll or Shuffle button.
3. Go to your DLNA device play control panel (example : Sonos App or Marantz App).
4. Click on the Next Track button and or anything related to request to play the next track from the play list.
5. Nothing happens.
6. The same command submitted from Jellyfin Web is working fine.

**Expected behavior**

When selecting to play the next track of the play list on the DLNA device, we expect the current track stops playing and the next one starts playing. If it is the end of the list, it should stop playing.

**Logs**

No error log has been noticed.

**Screenshots**

None

**Additional context**

None

### PR DESCRIPTION
**Pull request by ssenart**
**[DLNA] [NextTrack] [Gapless] Fix 5644 and 5294**

Fix 5644 : [DLNA] [Music] Next track command from any DLNA device does not do anything
Fix 5294 : gaps between tracks ... 

**Changes**

Implement the new DLNA message SetNextAVTransport() that permits to the renderer to prefetch the next track to play in the list.
It will have an effect if the renderer implement that feature :
- On my Sonos, it works perfectly.
- On my Marantz, I have to select 'replay once' option for it to work => strange.
- On JRiver Player, it works perfectly.
- On BubbleUPnP Android Player, it does not work at all : Possibly SetNextAVTransport() feature is not supported.

The other positive effect is that it removes the gap between tracks.

Feedbacks on other renderers are welcome !

**Issues**

Fixes #5644
Fixes #5294

### CODE REVIEW

**Comment by BaronGreenback** [comment]
Note to whomever: SetNextAVTransport isn't supported by all devices - so when mentioning gapless playback in any documentation should be qualified with "if supported".

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 1]
```diff
@@ -156,6 +172,10 @@ private async void OnDeviceMediaChanged(object sender, MediaChangedEventArgs e)
                 var newItemProgress = GetProgressInfo(streamInfo);
 
                 await _sessionManager.OnPlaybackStart(newItemProgress).ConfigureAwait(false);
+
+                // Send a message to the DLNA device to notify what is the next track in the playlist.
+                var currentItemIndex = _playlist.FindIndex(item => item.StreamInfo.ItemId == streamInfo.ItemId);
```
[the remark this answers is not in the document]
Comment (BaronGreenback) [reply Emby.Dlna/PlayTo/PlayToController.cs:177]: thought __currentPlaylistIndex_ contained the current item being played?

**Comment by BaronGreenback** [comment]
Not particular to this change, and would welcome comment as I may be over cautious - but shouldn't all playlist array lookups be behind some form of locks.

Only say this as it would be possible for the dlna devices trigger events which effect the list at the same time.
eg. OnDevicePlaybackStopped clears the playList - so theoretically cause an index out of bounds.

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 2]
```diff
@@ -736,6 +764,10 @@ private async Task SetAudioStreamIndex(int? newIndex)
 
                     await _device.SetAvTransport(newItem.StreamUrl, GetDlnaHeaders(newItem), newItem.Didl, CancellationToken.None).ConfigureAwait(false);
 
+                    // Send a message to the DLNA device to notify what is the next track in the play list.
+                    var newItemIndex = _playlist.FindIndex(item => item.StreamUrl == newItem.StreamUrl);
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/PlayToController.cs:768]: Isn't SeekAfterTransportChange unnecessary if SendNextTrackMessage is supported?

Comment (BaronGreenback) [reply Emby.Dlna/PlayTo/PlayToController.cs:768]: Also, do you know if dlna device, when the track changes, still sends a change notification?

File: Emby.Dlna/PlayTo/PlayToController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BaronGreenback) [reply Emby.Dlna/PlayTo/PlayToController.cs:177]: warning - it isn't set to -1 when the playlist is cleared in a couple of places.

**Comment by ssenart** [comment]
> Not particular to this change, and would welcome comment as I may be over cautious - but shouldn't all playlist array lookups be behind some form of locks.
> 
> Only say this as it would be possible for the dlna devices trigger events which effect the list at the same time.
> eg. OnDevicePlaybackStopped clears the playList - so theoretically cause an index out of bounds.

All the code here is coroutine oriented (async/await). I don't think there are multiple threads (I did not check). 

> Not particular to this change, and would welcome comment as I may be over cautious - but shouldn't all playlist array lookups be behind some form of locks.
> 
> Only say this as it would be possible for the dlna devices trigger events which effect the list at the same time.
> eg. OnDevicePlaybackStopped clears the playList - so theoretically cause an index out of bounds.

Since all is coroutine (async/await), should not all this code be executed by the same single thread ?

File: Emby.Dlna/PlayTo/PlayToController.cs (hunk 2 above)
Comment (ssenart) [reply Emby.Dlna/PlayTo/PlayToController.cs:768]: I don't know. In the code here, there is a timer that periodicaly polls the DLNA device status and synchronizes with it. That suggests there isn't any kind of callback notification (at the time this code was written).

Comment (BaronGreenback) [reply Emby.Dlna/PlayTo/PlayToController.cs:768]: Sorry - confusing this with the newer version which implements dlna callbacks. (The device triggers updates on changes.)

File: Emby.Dlna/PlayTo/PlayToController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BaronGreenback) [reply Emby.Dlna/PlayTo/PlayToController.cs:177]: there are a couple of places in the code where the playlist is reset (_playlist.Clear()), but _currentPlaylistIndex isn't reset.

Was hoping the plugin would have been ready yet - as it includes multiple fixed (including this code now) - _currentPlaylistIndex is one of them.

**Comment by BaronGreenback** [comment]
> > Not particular to this change, and would welcome comment as I may be over cautious - but shouldn't all playlist array lookups be behind some form of locks.
> > Only say this as it would be possible for the dlna devices trigger events which effect the list at the same time.
> > eg. OnDevicePlaybackStopped clears the playList - so theoretically cause an index out of bounds.
> 
> Since all is coroutine (async/await), should not all this code be executed by the same single thread ?

Again - confusing with the callbacks on the next version.

**Comment by ssenart** [comment]
I agree. SetNextAVTransport() support is mandatory on Device side. Otherwise, it behaves as before without the fix.

You're right, clicking next/next/next/next without waiting few seconds between may do nothing. The explanation is still with this Timer that resynchronize with the Renderer every 10 seconds (hardcoded). The click on next button is taking in account just at resynchronization time and a new SetNextAVTransport() message is then sent.

**Review by BaronGreenback — APPROVED** [review]

File: Emby.Dlna/PlayTo/Device.cs [hunk 3]
```diff
@@ -368,6 +368,42 @@ public async Task SetAvTransport(string url, string header, string metaData, Can
...
+         */
+        public async Task SetNextAvTransport(string url, string header, string metaData, CancellationToken cancellationToken = default)
+        {
+            var avCommands = await GetAVProtocolAsync(cancellationToken).ConfigureAwait(false);
+
+            url = url.Replace("&", "&amp;", StringComparison.Ordinal);
+
+            _logger.LogDebug("{0} - SetNextAvTransport Uri: {1} DlnaHeaders: {2}", Properties.Name, url, header);
+
+            var command = avCommands.ServiceActions.FirstOrDefault(c => c.Name == "SetNextAVTransportURI");
```
Comment (crobibero) [inline Emby.Dlna/PlayTo/Device.cs:383]:
```suggestion
            var command = avCommands.ServiceActions.FirstOrDefault(c => string.Equals(c.Name, "SetNextAVTransportURI", StringComparison.OrdinalIgnoreCase);
```

File: Emby.Dlna/PlayTo/Device.cs [hunk 4]
```diff
@@ -368,6 +368,42 @@ public async Task SetAvTransport(string url, string header, string metaData, Can
...
+         * SetNextAvTransport is used to specify to the DLNA device what is the next track to play.
+         * Without that information, the next track command on the device does not work.
+         */
+        public async Task SetNextAvTransport(string url, string header, string metaData, CancellationToken cancellationToken = default)
+        {
+            var avCommands = await GetAVProtocolAsync(cancellationToken).ConfigureAwait(false);
+
+            url = url.Replace("&", "&amp;", StringComparison.Ordinal);
+
+            _logger.LogDebug("{0} - SetNextAvTransport Uri: {1} DlnaHeaders: {2}", Properties.Name, url, header);
```
Comment (crobibero) [inline Emby.Dlna/PlayTo/Device.cs:381]:
```suggestion
            _logger.LogDebug("{PropertyName} - SetNextAvTransport Uri: {Url} DlnaHeaders: {Header}", Properties.Name, url, header);
```

**Review by crobibero — APPROVED** [review]

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 5]
```diff
@@ -623,6 +653,9 @@ private async Task SetPlaylistIndex(int index, CancellationToken cancellationTok
 
             await _device.SetAvTransport(currentitem.StreamUrl, GetDlnaHeaders(currentitem), currentitem.Didl, cancellationToken).ConfigureAwait(false);
 
+            // Send a message to the DLNA device to notify what is the next track in the play list.
+            await SendNextTrackMessage(index, CancellationToken.None);
```
Comment (cvium) [inline Emby.Dlna/PlayTo/PlayToController.cs:657]:
```suggestion
            await SendNextTrackMessage(index, cancellationToken);
```

**Review by cvium — APPROVED** [review]

File: Emby.Dlna/PlayTo/Device.cs [hunk 6]
```diff
@@ -368,6 +368,42 @@ public async Task SetAvTransport(string url, string header, string metaData, Can
...
+         */
+        public async Task SetNextAvTransport(string url, string header, string metaData, CancellationToken cancellationToken = default)
+        {
+            var avCommands = await GetAVProtocolAsync(cancellationToken).ConfigureAwait(false);
+
+            url = url.Replace("&", "&amp;", StringComparison.Ordinal);
+
+            _logger.LogDebug("{PropertyName} - SetNextAvTransport Uri: {Url} DlnaHeaders: {Header}", Properties.Name, url, header);
+
+            var command = avCommands.ServiceActions.FirstOrDefault(c => string.Equals(c.Name, "SetNextAVTransportURI", StringComparison.OrdinalIgnoreCase);
```
Comment (cvium) [inline Emby.Dlna/PlayTo/Device.cs:383]:
```suggestion
            var command = avCommands.ServiceActions.FirstOrDefault(c => string.Equals(c.Name, "SetNextAVTransportURI", StringComparison.OrdinalIgnoreCase));
```

**Pull request merged**